UI to import SEMs file and include in exported results

carolinemodic opened the pull request:
Depends on https://github.com/votingworks/vxsuite/pull/264
Adds a button to import an external SEMS file alongside the CVRs. Shows that file in the CVR file list, and (optionally) includes it in the results. 

CVR File List 
I put the name of the external file in the "scanner-id" column... this is a little weird, open to feedback/other ideas. 
I also moved the two remove buttons in a new row as 5 buttons felt like a lot in one row but I don't feel strongly.
![Screen Shot 2021-02-12 at 3 55 04 PM](https://user-images.githubusercontent.com/14897017/107836745-12335800-6d53-11eb-8cac-7c704e652dae.png)

Export Modal
I wasn't sure if we even want to give them the option not to include this data, so open to feedback/other ideas
![Screen Shot 2021-02-12 at 4 56 39 PM](https://user-images.githubusercontent.com/14897017/107836828-4c9cf500-6d53-11eb-96f9-dca8349ba5db.png)


Video of E2E Interactions:
(fyi I took this before removing the button for the combination screen but that is also now removed) 

https://user-images.githubusercontent.com/14897017/107836847-5b83a780-6d53-11eb-8805-d99e8dec6cf1.mov

beausmith reviewed (approved):
Looks good. Just a few suggestions.

FYI: Github breaks commit message into multiple lines when long:

```
feat(election-manager): import sems file alongside cvrs and inlcude in…

… export
```

I suggest shorter with colon after closing parentheses:
`feat(election-manager): Display and export SEMS results`

carolinemodic commented on apps/election-manager/src/components/ExportFinalResultsModal.tsx:
```diff
@@ -45,17 +42,32 @@ const ExportFinalResultsModal: React.FC<Props> = ({
+
+  const isTestMode = castVoteRecordFiles?.fileMode === 'test'
+
+  useEffect(() => {
+    setDefaultFilename(
+      generateFinalExportDefaultFilename(
+        isTestMode,
+        electionDefinition!.election
+      )
+    )
+  }, [isOpen])
```
I don't want it to create a new name every time you toggle the segmented button for including the SEMs data, since the timestamp on the file is shown in the dialog and is to the second it's kind of distracting/weird for that to be changing.

eventualbuddha replied:
Wouldn't this work without the `useEffect`?

```ts
const [defaultFilename, setDefaultFilename] = useState(      
  generateFinalExportDefaultFilename(
    isTestMode,
    electionDefinition!.election
  )
)
```

carolinemodic replied:
Since this is a model it gets rendered whenever the surrounding tallyscreen page is loaded, and I do want to regenerate the filename when the modal is closed and then reopened, which is why I'm depending on isOpen changing in the useEffect

beausmith replied:
Interesting issue. Just seems like useEffect isn't the "right" way to do this. I would have also suggested a solution similar to what Brian suggested. Maybe add a comment in the useEffect block to clarify what the useEffect is for?

eventualbuddha commented on apps/election-manager/.eslintrc.js:
```diff
@@ -60,6 +60,7 @@ module.exports = {
     'import/resolver': {
       node: {
         extensions: allExtensions,
+        moduleDirectory: ['node_modules', 'src/'],
```
Why?

carolinemodic replied:
Ah yeah I wanted to ask you about this, I was suddenly getting build errors about not being able to find lodash (even though it was installed) midway through working on this branch and this was the only way I could figure out how to fix it and get things building again but I don't actually understand what I'm doing here.

eventualbuddha commented:
> FYI: Github breaks commit message into multiple lines when long:
> 
> ```
> feat(election-manager): import sems file alongside cvrs and inlcude in…
> 
> … export
> ```
> 
> I suggest shorter with colon after closing parentheses:
> `feat(election-manager): Display and export SEMS results`

Also, typo in the commit message: `inlcude`.

carolinemodic commented on apps/election-manager/src/utils/file.ts:
```diff
@@ -8,3 +9,31 @@ export async function copyFile(inputFile: File): Promise<File> {
+
+export async function exportFileToStorageString(
+  inputFile: File,
+  fileContent?: string
```
No, I was thinking that would make sense in the future but in retrospect I shouldn't pre-optimize this.

carolinemodic commented:
@beausmith @eventualbuddha After thinking more about the structure of this and some of your comments I decided to pull the "File" for the externalVoteRecordsFile out of the Tally object and make it a top level item in AppContext. This more closely mirrors how the CastVoteRecords and Tallies are stored for our internal data. I made this a new refactor commit to make the changes easier to review but its probably large enough to be worth another pass. Note I also moved the useEffect block that computes the tallies from the files (for both this new one and existing CVRs) from TallyScreen to AppRoot so we're not randomly recomputing these tallies anytime you move on and off of the tally tab in election manager. 

The other changes to address comments I just squashed into the prior commit.

beausmith commented on apps/election-manager/src/AppRoot.tsx:
```diff
@@ -88,6 +98,11 @@ const AppRoot: React.FC<Props> = ({ storage }) => {
+  const [
+    fullElectionExternalTally,
+    setFullElectionExternalTally,
+  ] = useState<OptionalFullElectionExternalTally>(undefined)
```
Interesting. I wasn't aware that the type didn't have to include all possible types here.

eventualbuddha replied:
If you look at the type of `useState` you can see why:

```ts
    /**
     * Returns a stateful value, and a function to update it.
     *
     * @version 16.8.0
     * @see https://reactjs.org/docs/hooks-reference.html#usestate
     */
    function useState<S>(initialState: S | (() => S)): [S, Dispatch<SetStateAction<S>>];
    // convenience overload when first argument is omitted
    /**
     * Returns a stateful value, and a function to update it.
     *
     * @version 16.8.0
     * @see https://reactjs.org/docs/hooks-reference.html#usestate
     */
    function useState<S = undefined>(): [S | undefined, Dispatch<SetStateAction<S | undefined>>];
```

It's an overloaded function, and the second overload takes no arguments. So when you pass a type parameter like `FullElectionExternalTally` but no arguments, the second overload is matched. `S` becomes `FullElectionExternalTally` and the type of the return value is `[FullElectionExternalTally | undefined, Dispatch<SetStateAction<FullElectionExternalTally | undefined>>]`.

eventualbuddha commented on apps/election-manager/src/AppRoot.tsx:
```diff
@@ -163,10 +185,84 @@ const AppRoot: React.FC<Props> = ({ storage }) => {
+  const computeVoteCounts = useCallback(
+    async (castVoteRecords: CastVoteRecordLists) => {
+      setIsTabulationRunning(true)
+      const fullTally = await computeFullElectionTally(
```
This `await` doesn't do anything.
